File: src/battery/battery.cc
#include "battery/battery.h"

#include "battery/battery_impl.h"
#include "encos/driver_manager.h"
#include "protocol/route_ids.h"

namespace encos {

namespace {

uint16_t ReadU16Le(const uint8_t* data) {
    return static_cast<uint16_t>(static_cast<uint16_t>(data[0]) |
                                 (static_cast<uint16_t>(data[1]) << 8));
}

int16_t ReadI16Le(const uint8_t* data) {
    return static_cast<int16_t>(ReadU16Le(data));
}

void SetBit(uint8_t& target, uint8_t bit_index, bool enabled) {
    if (enabled) {
        target = static_cast<uint8_t>(target | (1u << bit_index));
    }
}

}  // namespace

bool Battery::Impl::OnMessage(const MotorPackMsg& message) {
    BatteryStatus snapshot{};
    std::function<void(const BatteryStatus&)> callback;
    bool status_updated = false;
    const auto now = std::chrono::steady_clock::now();
    const auto route_ids = protocol::BatteryStatusIds(idx);
    {
        platform::LockGuard<platform::Mutex> lock(status_mutex);
        if (message.id == route_ids[0] && message.len >= 8) {
            BatteryState parsed{};
            parsed.is_master = message.data[0] != 0;
            parsed.soc = message.data[1] / 100.0f;
            parsed.voltage = ReadU16Le(message.data + 2) * 0.01f;
            parsed.allowed_discharge_current = ReadU16Le(message.data + 4) * 0.00001f;
            parsed.allowed_charge_current = ReadU16Le(message.data + 6) * 0.00001f;
            current_status.state = parsed;
            last_state_update = now;
            status_updated = true;
        } else if (message.id == route_ids[1] && message.len >= 8) {
            BatteryTemp parsed{};
            parsed.battery = static_cast<float>(ReadI16Le(message.data));
            parsed.mos = static_cast<float>(ReadI16Le(message.data + 2));
            parsed.discharge_current = ReadU16Le(message.data + 4) * 0.00001f;
            parsed.charge_current = ReadU16Le(message.data + 6) * 0.00001f;
            current_status.temp = parsed;
            last_temp_update = now;
            status_updated = true;
        } else if (message.id == route_ids[2] && message.len >= 2) {
            BatteryError parsed{};
            parsed.could_not_charge = (message.data[0] & 0x01) != 0;
            parsed.could_not_discharge = (message.data[0] & 0x02) != 0;
            parsed.low_battery = (message.data[0] & 0x04) != 0;
            parsed.over_current_steady = (message.data[0] & 0x08) != 0;
            parsed.over_current_peak = (message.data[0] & 0x10) != 0;
            parsed.over_current_charge = (message.data[0] & 0x20) != 0;
            parsed.battery_over_temp = (message.data[0] & 0x40) != 0;
            parsed.mos_over_temp = (message.data[0] & 0x80) != 0;
            parsed.could_not_communicate = (message.data[1] & 0x01) != 0;
            parsed.stopped_emergency = (message.data[1] & 0x02) != 0;
            parsed.charger_fault = (message.data[1] & 0x04) != 0;
            current_status.error = parsed;
            last_error_update = now;
            status_updated = true;
        } else if (message.id == route_ids[3] && message.len >= 1) {
            BatteryActiveCommands parsed{};
            parsed.shutdown_request = (message.data[0] & 0x01) != 0;
            parsed.discharge_request = (message.data[0] & 0x02) != 0;
            parsed.force_shutdown_broadcast = (message.data[0] & 0x04) != 0;
            parsed.allow_charging = (message.data[0] & 0x08) != 0;
            parsed.fault_shutdown_broadcast = (message.data[0] & 0x10) != 0;
            parsed.mos_status = (message.data[0] & 0x20) != 0;
            current_status.active_commands = parsed;
            last_active_commands_update = now;
            status_updated = true;
        }
        if (status_updated) {
            snapshot = current_status;
            callback = on_status;
        }
    }

    if (status_updated && callback) {
        callback(snapshot);
    }
    return status_updated;
}

BatteryStatus Battery::Impl::GetStatusSnapshot() {
    platform::LockGuard<platform::Mutex> lock(status_mutex);
    auto now = std::chrono::steady_clock::now();
    if (now - last_state_update > state_timeout) {
        current_status.state.reset();
    }
    if (now - last_temp_update > state_timeout) {
        current_status.temp.reset();
    }
    if (now - last_active_commands_update > state_timeout) {
        current_status.active_commands.reset();
    }
    if (now - last_error_update > state_timeout) {
        current_status.error = {.comm_timeout = true};
    }
    return current_status;
}

void Battery::Impl::SendPassiveCommands(const BatteryPassiveCommands& commands) {
    platform::LockGuard<platform::Mutex> lock(command_mutex);
    MotorPackMsg msg{};
    msg.id = 0x4F4u + idx;
    msg.len = 2;
    SetBit(msg.data[0], 0, commands.allow_shutdown);
    SetBit(msg.data[0], 1, commands.allow_discharge);
    SetBit(msg.data[0], 2, commands.parallel_discharge);
    SetBit(msg.data[0], 3, commands.force_shutdown);
    SetBit(msg.data[0], 4, commands.request_charging);
    SetBit(msg.data[0], 5, commands.fault_shutdown_broadcast);
    SetBit(msg.data[0], 6, commands.configure_fault_thresholds);
    SetBit(msg.data[0], 7, commands.clear_fault);
    SetBit(msg.data[1], 0, commands.factory_mode);
    SetBit(msg.data[1], 1, commands.debug);
    writer(msg);
}

Battery::Battery(Bus* bus, uint16_t battery_idx, LoggerPtr logger,
                 std::function<void(const MotorPackMsg&)> writer) {
    impl_ = std::make_unique<Impl>();
    impl_->idx = battery_idx;
    impl_->bus = bus;
    impl_->writer = std::move(writer);
    impl_->logger_ = std::move(logger);
}

Battery::~Battery() = default;

BatteryStatus Battery::GetStatus() {
    auto operation = EncosDriverManager::Instance().AcquireDeviceOperation(this);
    return impl_->GetStatusSnapshot();
}

void Battery::OnMessage(const MotorPackMsg& message) {
    (void) impl_->OnMessage(message);
}

void Battery::SetOnStatus(std::function<void(const BatteryStatus&)> callback) {
    auto operation = EncosDriverManager::Instance().AcquireDeviceOperation(this);
    platform::LockGuard<platform::Mutex> lock(impl_->status_mutex);
    impl_->on_status = std::move(callback);
}

void Battery::SendPassiveCommands(const BatteryPassiveCommands& commands) {
    auto operation = EncosDriverManager::Instance().AcquireDeviceOperation(this);
    impl_->SendPassiveCommands(commands);
}

void Battery::ClearFault() {
    auto operation = EncosDriverManager::Instance().AcquireDeviceOperation(this);
    BatteryPassiveCommands commands{};
    commands.clear_fault = true;
    impl_->SendPassiveCommands(commands);
}

void Battery::RequestCharging(bool enabled) {
    auto operation = EncosDriverManager::Instance().AcquireDeviceOperation(this);
    BatteryPassiveCommands commands{};
    commands.request_charging = enabled;
    impl_->SendPassiveCommands(commands);
}

void Battery::AllowDischarge(bool enabled) {
    auto operation = EncosDriverManager::Instance().AcquireDeviceOperation(this);
    BatteryPassiveCommands commands{};
    commands.allow_discharge = enabled;
    impl_->SendPassiveCommands(commands);
}

}  // namespace encos

File: npm/src/battery.ts
import { assertOk } from './native.js'
import type { BusWrapper } from './bus.js'
import type { BatteryPassiveCommands, BatteryStatusResult } from './types.js'

export class BatteryWrapper {
  bus: BusWrapper
  runtime: BusWrapper['runtime']
  handle: number
  index: number
  disposed: boolean

  constructor(bus: BusWrapper, handle: number, index: number) {
    this.bus = bus
    this.runtime = bus.runtime
    this.handle = handle
    this.index = index
    this.disposed = false
  }

  getStatus(): BatteryStatusResult {
    this.#assertLive()
    const values = this.runtime.module._malloc(27 * 8)
    const meta = this.runtime.module._malloc(3 * 4)
    try {
      const code = this.runtime.calls.batteryGetStatus(this.handle, values, 27, meta, 3)
      assertOk(this.runtime, code, 'failed to get battery status')
      const hasState = this.runtime.module.getValue(meta, 'i32') !== 0
      const hasTemp = this.runtime.module.getValue(meta + 4, 'i32') !== 0
      const hasActiveCommands = this.runtime.module.getValue(meta + 8, 'i32') !== 0
      return {
        state: hasState
          ? {
              isMaster: this.#doubleAt(values, 0) !== 0,
              soc: this.#doubleAt(values, 1),
              voltage: this.#doubleAt(values, 2),
              allowedDischargeCurrent: this.#doubleAt(values, 3),
              allowedChargeCurrent: this.#doubleAt(values, 4),
            }
          : undefined,
        temp: hasTemp
          ? {
              battery: this.#doubleAt(values, 5),
              mos: this.#doubleAt(values, 6),
              dischargeCurrent: this.#doubleAt(values, 7),
              chargeCurrent: this.#doubleAt(values, 8),
            }
          : undefined,
        error: {
          couldNotCharge: this.#doubleAt(values, 9) !== 0,
          couldNotDischarge: this.#doubleAt(values, 10) !== 0,
          lowBattery: this.#doubleAt(values, 11) !== 0,
          overCurrentSteady: this.#doubleAt(values, 12) !== 0,
          overCurrentPeak: this.#doubleAt(values, 13) !== 0,
          overCurrentCharge: this.#doubleAt(values, 14) !== 0,
          batteryOverTemp: this.#doubleAt(values, 15) !== 0,
          mosOverTemp: this.#doubleAt(values, 16) !== 0,
          couldNotCommunicate: this.#doubleAt(values, 17) !== 0,
          stoppedEmergency: this.#doubleAt(values, 18) !== 0,
          chargerFault: this.#doubleAt(values, 19) !== 0,
          commTimeout: this.#doubleAt(values, 20) !== 0,
        },
        activeCommands: hasActiveCommands
          ? {
              shutdownRequest: this.#doubleAt(values, 21) !== 0,
              dischargeRequest: this.#doubleAt(values, 22) !== 0,
              forceShutdownBroadcast: this.#doubleAt(values, 23) !== 0,
              allowCharging: this.#doubleAt(values, 24) !== 0,
              faultShutdownBroadcast: this.#doubleAt(values, 25) !== 0,
              mosStatus: this.#doubleAt(values, 26) !== 0,
            }
          : undefined,
      }
    } finally {
      this.runtime.module._free(values)
      this.runtime.module._free(meta)
    }
  }

  sendPassiveCommands(commands: BatteryPassiveCommands): void {
    this.#assertLive()
    let byte0 = 0
    let byte1 = 0
    byte0 |= commands.allowShutdown ? 1 << 0 : 0
    byte0 |= commands.allowDischarge ? 1 << 1 : 0
    byte0 |= commands.parallelDischarge ? 1 << 2 : 0
    byte0 |= commands.forceShutdown ? 1 << 3 : 0
    byte0 |= commands.requestCharging ? 1 << 4 : 0
    byte0 |= commands.faultShutdownBroadcast ? 1 << 5 : 0
    byte0 |= commands.configureFaultThresholds ? 1 << 6 : 0
    byte0 |= commands.clearFault ? 1 << 7 : 0
    byte1 |= commands.factoryMode ? 1 << 0 : 0
    byte1 |= commands.debug ? 1 << 1 : 0
    const code = this.runtime.calls.batterySendPassiveCommands(this.handle, byte0, byte1)
    assertOk(this.runtime, code, 'failed to send battery passive commands')
  }

  clearFault(): void {
    this.#assertLive()
    const code = this.runtime.calls.batteryClearFault(this.handle)
    assertOk(this.runtime, code, 'failed to clear battery fault')
  }

  requestCharging(enabled: boolean): void {
    this.#assertLive()
    const code = this.runtime.calls.batteryRequestCharging(this.handle, enabled ? 1 : 0)
    assertOk(this.runtime, code, 'failed to set battery charging request')
  }

  allowDischarge(enabled: boolean): void {
    this.#assertLive()
    const code = this.runtime.calls.batteryAllowDischarge(this.handle, enabled ? 1 : 0)
    assertOk(this.runtime, code, 'failed to set battery discharge permission')
  }

  #assertLive(): void {
    this.bus.adapter._assertLiveForChild()
    if (this.disposed || this.bus.disposed) {
      throw new Error('battery is disposed')
    }
  }

  #doubleAt(ptr: number, index: number): number {
    return this.runtime.module.getValue(ptr + index * 8, 'double')
  }
}
